payments/providers/cpu_ceepos.py

```diff
@@ -0,0 +1,361 @@
+from requests.exceptions import RequestException
+
+from ..exceptions import (
+    DuplicateOrderError,
+    OrderStateTransitionError,
+    PayloadValidationError,
+    PaymentCreationFailedError,
+    ServiceUnavailableError,
+    UnknownReturnCodeError,
+)
+from ..models import Order, OrderLine
+from ..utils import price_as_sub_units
+from .base import PaymentProvider
+
+LOG = logging.getLogger(__name__)
+
+# Keys the provider expects to find in the config
+RESPA_PAYMENTS_CEEPOS_API_URL = "RESPA_PAYMENTS_CEEPOS_API_URL"
+RESPA_PAYMENTS_CEEPOS_API_KEY = "RESPA_PAYMENTS_CEEPOS_API_KEY"
+RESPA_PAYMENTS_CEEPOS_API_SECRET = "RESPA_PAYMENTS_CEEPOS_API_SECRET"
+
+
+class CPUCeeposProvider(PaymentProvider):
+    """CPU Ceepos specific integration utilities and configuration"""
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.url_payment_api = self.config.get(RESPA_PAYMENTS_CEEPOS_API_URL)
+
+    @staticmethod
+    def get_config_template() -> Dict[str, str]:
+        """Keys and values that CeePos requires from environment"""
+        return {
+            RESPA_PAYMENTS_CEEPOS_API_URL: str,
+            RESPA_PAYMENTS_CEEPOS_API_KEY: str,
+            RESPA_PAYMENTS_CEEPOS_API_SECRET: str,
+        }
+
+    def initiate_payment(self, order: Order) -> str:
+        """
+        Creates a payment to the provider. The insertion order of the
+        fields in the payload data is important here since the values
+        are used for checksum calculations.
+
+        Returns an URL to which the user is redirected
+        to actually pay the order.
+        """
+
+        payload = {
+            "ApiVersion": "3.0.0",
+            "Source": self.config.get(RESPA_PAYMENTS_CEEPOS_API_KEY),
+            "Id": str(order.order_number),
+            "Mode": 3,
+            "Action": "new payment",
+        }
+
+        self.payload_add_products(payload, order)
+        self.payload_add_customer(payload, order)
+        self.payload_add_return_and_notification_urls(payload)
+        self.payload_add_checksum(payload)
+
+        try:
+            r = requests.post(self.url_payment_api, json=payload, timeout=60)
+            r.raise_for_status()
+            return self.handle_initiate_payment_response(r.json())
+        except RequestException as e:
+            raise ServiceUnavailableError("Payment service is unreachable") from e
+
+    def payload_add_products(self, payload, order):
+        """Attaches product data to the payload
+
+        Order lines that contain bought products are retrieved through order"""
+        reservation = order.reservation
+        order_lines = OrderLine.objects.filter(order=order.id)
+        items = []
+        for order_line in order_lines:
+            product = order_line.product
+            int_tax = int(product.tax_percentage)
+            assert (
+                int_tax == product.tax_percentage
+            )  # make sure the tax is a whole number
```

Comment:
So this will just break if the government decides to raise the VAT by 0.5%? Maybe that's OK, but I think it shouldn't accept even all integer values.

I think this should rather be something like a dictionary than bare conversion to integer. Especially, since the taxcode should be a string according to the API spec.

Maybe there should be a function like this instead:
```python
def get_taxcode(product: Product) -> str:
    tax = product.tax_percentage
    taxcode = {
        24_000_000: "24",
        14_000_000: "14",
        10_000_000: "10",
        0: "0",
    }.get(int(tax * 1_000_000))
    if not taxcode:
        raise ValueError(f"Unsupported tax percentage {tax} for product {product}")
    return taxcode
```

The dict data is from the API spec:
![image](https://user-images.githubusercontent.com/3531281/104700015-3aac4000-571c-11eb-8bb4-dfd0b44281c7.png)

However, it says that these tax codes are available in the test environment. Do you know if the same should be available for the production env too?

Reply:
Yes, those same tax classes will be available in the production env too. I think the dictionary-based handling of this sounds like a good idea. I will make the change.

Reply:
I added the function.